### ISSUE TITLE
Can't start - OpenCL device not found.

### ISSUE THREAD

**Issue by ShockWave2048** [virex-84/LLMGPT2#1]
Hi virex-84!
I can't run application. What could be the problem? 
(Win11, RTX3090, GPU-Z -> OpenCL 3.0, Cuda 12.8)
I tried switching ILGPU to the CUDA context (Help from Discord ILGPU), but it crashes when trying to calculate (create a new model)
Thanks for help or answer!

<img width="1934" height="1204" alt="Image" src="https://github.com/user-attachments/assets/bf757824-29b6-4df1-a386-812e24894f95" />

<img width="797" height="308" alt="Image" src="https://github.com/user-attachments/assets/547dfd14-43f2-41a3-be78-74c6eaa43ac4" />

**Comment by virex-84** [comment]
Added selection of all ILGPU 1.5.3 supported devices ( https://github.com/virex-84/LLMGPT2/commit/95870ea25d9b786a76b1108be71dedcd465ab45e ): CPU, OpenCL, CUDA. 

Try to select all available ones one by one. My configuration: AMD Ryzen AI 370 hx CPU with integrated AMD Radeon 890M graphics card. I physically don’t have any other hardware to test the application.

<img width="466" height="172" alt="Image" src="https://github.com/user-attachments/assets/178c6156-503c-4e2c-a6f1-54765c95d880" />

**Comment by ShockWave2048** [comment]
Yes, it shows CUDA device, but it crashes when starting model training.

<img width="486" height="121" alt="Image" src="https://github.com/user-attachments/assets/4953ed43-def8-4112-a3ea-33d350838ce0" />
<img width="1488" height="1225" alt="Image" src="https://github.com/user-attachments/assets/461c1032-79b8-4aa1-9b93-9aa7171c5aae" />

**Comment by virex-84** [comment]
I sent the change to MatrixOps, check it again.

**Comment by ShockWave2048** [comment]
I asked Gemini, and he advised me to create a separate accelerator for CUDA. It worked. I saved the GGUF model.

<img width="1629" height="1165" alt="Image" src="https://github.com/user-attachments/assets/ca388ee9-11f4-410f-a6b4-299c1de1e68a" />

**Comment by ShockWave2048** [comment]
Почему в ЛМ Студии может не понимать кодировку?

<img width="1165" height="948" alt="Image" src="https://github.com/user-attachments/assets/81c94787-19c9-484f-b858-cf018bdcaa36" />

**Comment by virex-84** [comment]
> I asked Gemini, and he advised me to create a separate accelerator for CUDA. It worked. I saved the GGUF model.

В любом случае в кернелах использовался Math, а должен был XMath. В MatrixOps эта проблема исправлена.


> Почему в ЛМ Студии может не понимать кодировку?

Необходимо забрать измененный MatrixOps, и переобучить. Обучать необходимо до Loss=0.12, чем меньше тем лучше. Недообученная модель выдает мусор.
Так же прежде чем экспортировать в gguf, можно в самой программе выбрать "4. Интерактивный чат" и проверить что выдает модель.

В статье есть описание: https://habr.com/ru/articles/1017484/

> Обучение модели:
> 
> Pretrain около 100 эпох с LR (learning rate) = 0.001
> 
> Tune (обучение диалогам) около 100 эпох с LR=0.0005 (loss упадет с 8,4 до 0,3), затем еще несколько раз с меньшим LR но больше эпох, пока loss не достигнет 0.12. Это самое долгое.

**Comment by ShockWave2048** [comment]
Огромное спасибо за ответы! Буду пробовать дальше. 
А Loss=0.12 и LR (learning rate) уже выставлен или там надо где то выставлять?

**Comment by virex-84** [comment]
LR (Learning rate) выдается по умолчанию, его можно изменять. Он как рычаг: чем больше - тем быстрее модель может обучатся на первых порах, но может и застрять в обучении (Loss не будет падать). Чем меньше - тем аккуратнее, тем меньшими шагами делает обучение, но учится долго. Тут нужен баланс. Уменьшают уже ближе к концу обучения.

Loss - это маркер того как модель обучилась: чем меньше - тем лучше.

пример в консоли:

> ═══ FINETUNE ═══
> Примеров: 1452
> 
>   Количество эпох (по умолчанию 27):       <- я ничего не ввёл и сразу нажал Enter, по умолчанию будет 27 эпох
>   Learning rate (по умолчанию 0,0001): 0.005      <- тут я ввёл 0.005 и нажал Enter. Было 0.0001, а я указал 0.005

**Comment by ShockWave2048** [comment]
Спасибо за советы! Все работает!
